docs/security-openid-connect.md

```diff
@@ -0,0 +1,224 @@
+
+# Optional settings (values shown are the defaults)
+# The path to the file containing the trusted certificate(s) of the token issuer(s). If not set, uses the default
+# trust store of the JVM.
+PULSAR_PREFIX_openIDTokenIssuerTrustCertsFilePath=
+# The JWT's claim to use for the role/principal during authorization.
+PULSAR_PREFIX_openIDRoleClaim=sub
+# The leeway, in seconds, to use when validating the token's expiration time.
+PULSAR_PREFIX_openIDAcceptedTimeLeewaySeconds=0
+
+# Cache settings
+PULSAR_PREFIX_openIDCacheSize=5
+PULSAR_PREFIX_openIDCacheRefreshAfterWriteSeconds=64800
+PULSAR_PREFIX_openIDCacheExpirationSeconds=86400
+PULSAR_PREFIX_openIDHttpConnectionTimeoutMillis=10000
+PULSAR_PREFIX_openIDHttpReadTimeoutMillis=10000
+
+# Whether to require that issuers use HTTPS. It is part of the OIDC spec to use HTTPS, so the default is true.
+# This setting is for testing purposes and is not recommended for any production environment.
+PULSAR_PREFIX_openIDRequireIssuersUseHttps=true
+
+# A setting describing how to handle discovery of the OpenID Connect configuration document when the issuer is not
+# in the list of allowed issuers. This setting is documented below.
+PULSAR_PREFIX_openIDFallbackDiscoveryMode=DISABLED
+```
+
+::: note
+When using OIDC for a client connecting through the proxy to the broker, it is necessary to set the broker's `openIDAcceptedTimeLeewaySeconds` to double the proxy's `authenticationRefreshCheckSeconds` configuration because the proxy caches the client's token and only refreshes it when it is expired. As such, in certain cases, when the proxy initiates a new connection to the broker, the token may not yet be expired in the proxy, but may be expired when it reaches the broker. You can mitigate this edge case by setting the broker's `openIDAcceptedTimeLeewaySeconds` correctly.
+:::
+
+## Enable OpenID Connect Authentication in the Function Worker
+
+To configure the Pulsar Function Worker to authenticate clients using OpenID Connect, add the following parameters to the `conf/functions_worker.yml` file. The documentation for these settings is [above](#enable-openid-connect-authentication-in-the-brokers-proxies-and-websocket-proxies).
+
+```yaml
+# Configuration to enable authentication
+authenticationEnabled: true
+authenticationProviders: "org.apache.pulsar.broker.authentication.oidc.AuthenticationProviderOpenID"
+properties:
+  openIDAllowedTokenIssuers: "https://my-issuer-1.com,https://my-issuer-2.com"
+  openIDAllowedAudiences: "audience-1,audience-2"
+  openIDTokenIssuerTrustCertsFilePath: ""
+  openIDRoleClaim: "sub"
+  openIDAcceptedTimeLeewaySeconds: 0
+  openIDCacheSize: 5
+  openIDCacheRefreshAfterWriteSeconds: 64800
+  openIDCacheExpirationSeconds: 86400
+  openIDHttpConnectionTimeoutMillis: 10000
+  openIDHttpReadTimeoutMillis: 10000
+  openIDRequireIssuersUseHttps: true
+  openIDFallbackDiscoveryMode: "DISABLED"
+```
+
+## OpenID Connect Kubernetes Authentication Flow
+
+Kubernetes has a---
```

Comment:
Looks like an unexpected copy&paste from row94 to row184?

Reply:
I'm not sure how I did that, but yes, that was an accidental copy and paste. Thanks for catching that!